src/framework/audio/internal/platform/jack/jackaudiodriver.cpp

```diff
@@ -20,279 +20,699 @@
+            muse::midi::Event e;
+            while (1) {
+                if (state->m_midiQueue.pop(e)) {
+                    sendEvent(e, pb);
+                } else {
+                    break;
+                }
+            }
+        }
+    } else {
+        muse::midi::Event e;
+        while (1) {
+            if (state->m_midiQueue.pop(e)) {
+                LOGW() << "no jack-midi-outport, consumed unused Event: " << e.to_string();
+            } else {
+                break;
+            }
+        }
+    }
+
+    if (!state->m_midiInputPorts.empty()) {
+        jack_port_t* port = state->m_midiInputPorts.front();
+        if (port) {
+            int segmentSize = jack_get_buffer_size(client);
+            void* pb = jack_port_get_buffer(port, segmentSize);
+            if (pb) {
+                muse::midi::Event ev;
+                jack_nframes_t n = jack_midi_get_event_count(pb);
+                for (jack_nframes_t i = 0; i < n; ++i) {
+                    jack_midi_event_t event;
+                    if (jack_midi_event_get(&event, pb, i) != 0) {
+                        continue;
+                    }
+                    int type = event.buffer[0];
+                    uint32_t data = 0;
+                    if ((type & 0xf0) == 0x90
+                        || (type & 0xf0) == 0x90) {
+                        data = 0x90
+                               | (type & 0x0f)
+                               | ((event.buffer[1] & 0x7F) << 8)
+                               | ((event.buffer[2] & 0x7F) << 16);
+                        Event e = Event::fromMIDI10Package(data);
+                        e = e.toMIDI20();
+                        if (e) {
+                            LOGI("-- jack midi-input-port send %i,%i,%i",
+                                 event.buffer[1],
+                                 event.buffer[2],
+                                 event.buffer[0]);
+                            state->m_eventReceived.send(static_cast<tick_t>(0), e);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return 0;
 }
 
-AudioDeviceID JackAudioDriver::outputDevice() const
+static int handle_buffersize_change(jack_nframes_t nframes, void* arg)
 {
-    return m_deviceId;
+    g_nframes = nframes;
+    return 0; // successfully reallocated buffer
 }
 
-bool JackAudioDriver::selectOutputDevice(const AudioDeviceID& deviceId)
+static void jack_cleanup_callback(void*)
 {
-    if (m_deviceId == deviceId) {
-        return true;
-    }
+}
+}
 
-    bool reopen = isOpened();
-    close();
-    m_deviceId = deviceId;
+JackDriverState::JackDriverState(mu::playback::IPlaybackController* playbackController)
+{
+    m_playbackController = playbackController;
+    m_deviceId = JACK_DEFAULT_DEVICE_ID;
+    m_deviceName = JACK_DEFAULT_IDENTIFY_AS;
 
-    bool ok = true;
-    if (reopen) {
-        ok = open(s_format2, &s_format2);
-    }
+    s_playbackController = playbackController;
```

Comment:
I see the driver's dependence on the controller. A driver is an independent thing and should not be dependent on anyone. Perhaps you should solve your problem through channels and the controller should subscribe to the channel and do something.

Reply:
> A driver is an independent thing and should not be dependent on anyone.

That has been addressed in two commits not part of this PR yet (because they are causing problem with syncing).
I've just pushed them now for completeness sake.